e2e-polybft/framework/test-cluster.go

```diff
@@ -554,3 +593,207 @@ func (c *TestCluster) InitSecrets(prefix string, count int) ([]types.Address, er
 
 	return result, nil
 }
+
+var (
+	defaultGasPrice = uint64(1879048192) // 0x70000000
+	defaultGasLimit = uint64(5242880)    // 0x500000
+)
+
+func (c *TestCluster) ExistsCode(t *testing.T, addr ethgo.Address) bool {
+	t.Helper()
+
+	client, err := jsonrpc.NewClient(c.Servers[0].JSONRPCAddr())
+	require.NoError(t, err)
+
+	code, err := client.Eth().GetCode(addr, ethgo.Latest)
+	if err != nil {
+		return false
+	}
+
+	if code == "0x" {
+		return false
+	}
+
+	return true
+}
+
+func (c *TestCluster) Call(t *testing.T, to types.Address, method *abi.Method,
+	args ...interface{}) map[string]interface{} {
+	t.Helper()
+
+	client, err := jsonrpc.NewClient(c.Servers[0].JSONRPCAddr())
+	require.NoError(t, err)
+
+	input, err := method.Encode(args)
+	require.NoError(t, err)
+
+	toAddr := ethgo.Address(to)
+
+	msg := &ethgo.CallMsg{
+		To:   &toAddr,
+		Data: input,
+	}
+	resp, err := client.Eth().Call(msg, ethgo.Latest)
+	require.NoError(t, err)
+
+	data, err := hex.DecodeString(resp[2:])
+	require.NoError(t, err)
+
+	output, err := method.Decode(data)
+	require.NoError(t, err)
+
+	return output
+}
+
+type TestCall struct {
+}
+
+func (c *TestCluster) Deploy(t *testing.T, sender ethgo.Key, bytecode []byte) *TestTxn {
+	t.Helper()
+
+	return c.SendTxn(t, sender, &ethgo.Transaction{Input: bytecode})
+}
+
+func (c *TestCluster) Transfer(t *testing.T, sender ethgo.Key, target types.Address, value *big.Int) *TestTxn {
+	t.Helper()
+
+	targetAddr := ethgo.Address(target)
+
+	return c.SendTxn(t, sender, &ethgo.Transaction{To: &targetAddr, Value: value})
+}
+
+func (c *TestCluster) MethodTxn(t *testing.T, sender ethgo.Key, target types.Address, input []byte) *TestTxn {
+	t.Helper()
+
+	targetAddr := ethgo.Address(target)
+
+	return c.SendTxn(t, sender, &ethgo.Transaction{To: &targetAddr, Input: input})
+}
+
+// SendTxn sends a transaction
+func (c *TestCluster) SendTxn(t *testing.T, sender ethgo.Key, txn *ethgo.Transaction) *TestTxn {
```

Comment:
You can use the transaction relayer as well in this function, and remove bunch of code.

Reply:
I commented out on this [here](https://github.com/0xPolygon/polygon-edge/pull/1306#discussion_r1139882067)

Reply:
Yeah, but the code for creating transaction, getting nonce, signing and sending is practically the same. I'm maybe missing the point, but I am not sure what else could we need there. If you wanted to make some custom processing of a transaction execution result, then I would say we can just let the `relayer` return the result, and then have a custom function to process the result. For waiting of the transaction receipt, we have that `WithReceiptTimeout` config function that allows us to configure a larger duration to wait for the receipt on the `relayer`.

Reply:
Yep, so initially, I was doing a bit more magic on the `TestTxn` type, for example:
```
.SendTxn().
  OnFinished(func() error { // custom on finish logic }).
  Wait()
```
but you are right that now it does not have any of those methods. Lemme see how many things I can move to the relayer.

Reply:
I had to remember a little bit about why I did this. So the `txrelayer` was not flexible enough for some of the patterns I had in mind (i.e send a specific transaction that you expect to fail) so this is how it evolved.

I think some (if not all) of the patterns can be moved to the `txrelayer` but I wanted to avoid changing code in other places out of the scope of this PR. I propose to have a followup PR to extend the usage of this framework to other unit tests (i.e. txpool e2e test would be sliced in half) to gather a bit of requirements and then move everything to `txrelayer`.

Reply:
Agreed.